### PR TITLE
Application Version Compliance Report

### PR DESCRIPTION
**Pull request by jingcheng16**
## Product Description

Frontend
image

Report
image

## Technical Summary

Ticket: https://dimagi.atlassian.net/browse/SAAS-16311

Review by commit. ( The commit history is basically how I improved the performance step by step )

This PR is to add a new Application Version Compliance Tile to the Enterprise Console.
This tile will iterate through all domain in this enterprise account, in each domain, iterate through all mobile workers, for each mobile worker, iterate through all their `last_build` in `last_builds`, (`last_builds` include the last build of the application they installed),  and see if the build is out of date, if it is, then append it to the report's row. 

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/enterprise/enterprise.py [hunk 1]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+            .source([
+                'username',
+                'reporting_metadata.last_builds',
+            ]))
+        for user in user_query.run().hits:
+            last_builds = user.get('reporting_metadata', {}).get('last_builds', [])
+            for build in last_builds:
+                app_id = build.get('app_id')
+                build_version = build.get('build_version')
+                if app_id not in apps or not build_version:
```
Comment (nospame) [inline corehq/apps/enterprise/enterprise.py:722]: Since this is iterating by domain, could a user have an app that's not in the passed in `apps`? A user here belongs to a domain, right, and should only have apps that belong to that same domain? I do see that this would also catch the situation where `app_id` is `None`.

File: corehq/apps/enterprise/enterprise.py [hunk 2]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+            self.builds_by_app_id[app_id] = app_es.run().hits
+
+        return self.builds_by_app_id[app_id]
+
+    def get_latest_build_version_at_time(self, all_builds, time):
+        """
+        Get the latest build version at the time
+
+        :param all_builds: List of raw build documents sorted by version in descending order
+        :param time: The date of the build version to compare against
```
Comment (nospame) [inline corehq/apps/enterprise/enterprise.py:754]: Might be worth specifying what date format is expected here. Seems like it's whatever is returned by `DateTimeProperty.deserialize`.

File: corehq/apps/enterprise/enterprise.py [hunk 3]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+                app_id = build.get('app_id')
+                build_version = build.get('build_version')
+                if app_id not in apps or not build_version:
+                    continue
+                build_version_date = DateTimeProperty.deserialize(build.get('build_version_date'))
+                all_builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+                latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                if not latest_version:
+                    all_builds = self.get_app_builds(domain, app_id)
+                    latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
```
Comment (nospame) [inline corehq/apps/enterprise/enterprise.py:729]: It seems like `get_app_builds` is only used by `get_latest_build_version_at_time`. Rather than just retrying the query here, could that `if not` logic be moved (or removed) if `get_app_builds` is called directly by `get_latest_build_version_at_time`?

File: corehq/apps/enterprise/enterprise.py [hunk 4]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+                    'last_released': DateTimeProperty.deserialize(released_date)
+                }
+                self.build_by_build_id[build_id] = build_info
+
+            if build_info['last_released'] <= time:
+                return build_info['version']
+        return None
+
+
+def _format_percentage_for_enterprise_tile(dividend, divisor):
```
Comment (nospame) [inline corehq/apps/enterprise/enterprise.py:778]: Is it correct to have this marked as "private" with the leading `_`? It seems like it's used in a few different places in this file now.

**Review by nospame — COMMENTED** [review]
I think this looks good! Did leave some context/structural/reasoning questions.

File: corehq/apps/enterprise/enterprise.py [hunk 5]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+        self.build_by_build_id = {}
+
+    @property
+    def headers(self):
+        return [_('Mobile Worker'), _('Project Space'), _('Application'),
+                _('Latest Version Available at Submission'),
+                _('Version in Use'), _('Last Used [UTC]')]
+
+    @property
+    def rows(self):
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:656]: Yes, the parent `EnterpriseReport` class's `rows` is iterate through all the domain object, to get all the domain object, [it iterate through all domain's name and get the domain object.](https://github.com/dimagi/commcare-hq/blob/848cf09907dadc7919d3727a602d50fb4c5c5b50/corehq/apps/enterprise/enterprise.py#L133-L134), it takes time for large enterprise account have more than 400 domains. Since this tile, only domain name is required, I prefer don't call `domains`.

File: corehq/apps/enterprise/enterprise.py [hunk 6]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+        for domain in self.account.get_domains():
+            rows.extend(self.rows_for_domain(domain))
+        return rows
+
+    @property
+    def total(self):
+        total_out_of_date_rows = 0
+        total_last_builds = 0
+
+        for domain in self.account.get_domains():
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:667]: This is done intentionally to avoid get domain objects for each domain

File: corehq/apps/enterprise/enterprise.py [hunk 7]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+    def _get_user_builds(self, domain, apps):
+        user_query = (UserES()
+            .domain(domain)
+            .mobile_users()
+            .source([
+                'username',
+                'reporting_metadata.last_builds',
+            ]))
+        for user in user_query.run().hits:
+            last_builds = user.get('reporting_metadata', {}).get('last_builds', [])
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:718]: I'm not sure. I think I'm just doing it in case it is missing. I saw we're doing it [here](https://github.com/dimagi/commcare-hq/blob/848cf09907dadc7919d3727a602d50fb4c5c5b50/corehq/apps/reports/standard/deployments.py#L336) for `reporting_metadata` as well. There is no harm to add a fallback value.

File: corehq/apps/enterprise/enterprise.py (hunk 1 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:722]: Iteration will skip when `last_builds` have build information for a deleted app.

File: corehq/apps/enterprise/enterprise.py (hunk 4 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:778]: My understanding of private function is it will only be used in the file, never be imported. So use it multiple times in one file seems fine?

File: corehq/apps/enterprise/enterprise.py (hunk 3 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:729]: Can you elaborate on this? Here what I'm trying to do is first get the latest 10 builds, and see if we can identify the latest version at use time from this 10 builds, if not, then query again to get all builds, and try to identify the latest version from all builds

File: corehq/apps/enterprise/enterprise.py (hunk 2 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:754]: I thought date by default means a datetime object, rather than string?

File: corehq/apps/enterprise/enterprise.py [hunk 8]
```diff
@@ -635,3 +639,87 @@ def rows_for_domain(self, domain_obj):
         if domain_obj.two_factor_auth:
             return []
         return [(domain_obj.name,)]
+
+
+class EnterpriseAppVersionComplianceReport(EnterpriseReport):
+    title = gettext_lazy('Application Version Compliance')
+    total_description = gettext_lazy('% of Applications Up-to-Date Across All Mobile Workers')
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: Thank you!! b024957ecb43d0261689e096d0e669e7b6566c21

File: corehq/apps/enterprise/enterprise.py [hunk 9]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+    def __init__(self, account, couch_user):
+        super().__init__(account, couch_user)
+        self.builds_by_app_id = {}
+        self.build_by_build_id = {}
+
+    @property
+    def headers(self):
+        return [_('Mobile Worker'), _('Project Space'), _('Application'),
+                _('Latest Version Available at Submission'),
+                _('Version in Use'), _('Last Used [UTC]')]
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: Didn't commit suggestion so I can combine this commit with the text update: b024957ecb43d0261689e096d0e669e7b6566c21

File: corehq/apps/enterprise/enterprise.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/enterprise/enterprise.py:661]: Got it, I missed that. Looks good then.

File: corehq/apps/enterprise/enterprise.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/enterprise/enterprise.py:723]: I'm not sure I agree in principle. To me, adding a fallback value indicates that _there is definitely_ a situation where the key can be missing, versus considering it an error state where something has gone wrong. I'd call this a non-blocking comment, though.

File: corehq/apps/enterprise/enterprise.py (hunk 1 above)
Comment (nospame) [reply corehq/apps/enterprise/enterprise.py:727]: Forgot about deleted apps. Makes sense to me, then!

File: corehq/apps/enterprise/enterprise.py (hunk 2 above)
Comment (nospame) [reply corehq/apps/enterprise/enterprise.py:759]: It's getting called with `build_version_date` that's comes from `DateTimeProperty.deserialize(build.get('build_version_date'))`. Is that a datetime object? I would prefer to see it specified in the docstring whether it is or not, so there's no ambiguity.

File: corehq/apps/enterprise/enterprise.py (hunk 4 above)
Comment (nospame) [reply corehq/apps/enterprise/enterprise.py:783]: Okay, I read up a little on this and I think you're correct!

File: corehq/apps/enterprise/enterprise.py (hunk 3 above)
Comment (nospame) [reply corehq/apps/enterprise/enterprise.py:734]: It makes sense what this is trying to do, I think what I'm getting at is that I think the retry logic should be handled in the method doing the fetching, and doesn't need to be seen by `_get_user_builds`. Maybe I'm overcomplicating it/trying to overabstract things. But I'm imagining something like:
```python
def get_latest_build_version_at_time(domain, app_id, time):
    builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
    latest_build = self._get_latest_build_version_if_present(builds, time)

    if latest_build is None:
        builds = self.get_app_builds(domain, app_id, start=self.INITIAL_QUERY_LIMIT)
        latest_build = self._get_latest_build_version_if_present(builds, time)

    return latest_build
```
Where `_get_latest_build_version_if_present` (or whatever naming) is handling the loop and returning `None` if the latest isn't in the builds that were passed in, and `get_app_builds` is modified to take a `start` parameter or something similar, so it doesn't return the same initial 10 builds along with the rest of them. Then `get_latest_build_version_at_time` will always return the actual latest build version.

Comment (gherceg) [reply corehq/apps/enterprise/enterprise.py:734]: +1 to separating out that logic. A helper method should be capable of doing everything to get the latest build version.

File: corehq/apps/enterprise/enterprise.py [hunk 10]
```diff
@@ -635,3 +634,153 @@ def rows_for_domain(self, domain_obj):
...
+            last_builds = user.get('reporting_metadata', {}).get('last_builds', [])
+            for build in last_builds:
+                app_id = build.get('app_id')
+                build_version = build.get('build_version')
+                if app_id not in apps or not build_version:
+                    continue
+                build_version_date = DateTimeProperty.deserialize(build.get('build_version_date'))
+                all_builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+                latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                if not latest_version:
```
Comment (gherceg) [inline corehq/apps/enterprise/enterprise.py:732]: It seems unfortunate that we have to fetch all builds to find the latest version available at the time of use. Is it possible to have a more targeted query using the built_on and/or last_released properties? The downside is that would probably be less effective for the current caching mechanism and in turn might result in more ES queries which could be worse, even if they are targeted. Is that something you have explored?

File: corehq/apps/enterprise/enterprise.py [hunk 11]
```diff
@@ -639,6 +639,7 @@ def rows_for_domain(self, domain_obj):
 class EnterpriseAppVersionComplianceReport(EnterpriseReport):
     title = gettext_lazy('Application Version Compliance')
     total_description = gettext_lazy('% of Applications Up-to-Date Across All Mobile Workers')
+    INITIAL_QUERY_LIMIT = 10
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:642]: I think normally the version should be not too out-of-date, so I thought 10 might be a good number...

File: corehq/apps/enterprise/enterprise.py (hunk 3 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: Refactored in 25e5e521275eec772dda66abc24667216cda82c6, also updated the docstring!

File: corehq/apps/enterprise/enterprise.py (hunk 2 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: Docstring updated with the refactoring: 25e5e521275eec772dda66abc24667216cda82c6

File: corehq/apps/enterprise/enterprise.py [hunk 12]
```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
...
+                build_version = build.get('build_version')
+                if app_id not in apps or not build_version:
+                    continue
+                build_version_date = DateTimeProperty.deserialize(build.get('build_version_date'))
+                all_builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+                latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                if not latest_version:
+                    all_builds = self.get_app_builds(domain, app_id)
+                    latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                yield user, build, latest_version
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:731]: _get_user_builds is also used in https://github.com/dimagi/commcare-hq/blob/25e5e521275eec772dda66abc24667216cda82c6/corehq/apps/enterprise/enterprise.py#L686-L698. I need those info for each row. Maybe I can return username instead of user?

File: corehq/apps/enterprise/enterprise.py [hunk 13]
```diff
@@ -635,3 +634,153 @@ def rows_for_domain(self, domain_obj):
...
+                ])
+
+        return rows
+
+    def total_for_domain(self, domain):
+        apps = get_app_ids_in_domain(domain)
+        total_last_builds = 0
+        total_out_of_date = 0
+
+        for user, build, latest_version in self._get_user_builds(domain, apps):
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: c446ad1ebfcd3440147dc3172f9a330afa07936d

File: corehq/apps/enterprise/enterprise.py (hunk 10 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: A more targeted query takes 200 sec in total...I don't understand why it takes that long...
```
app_es = (
    AppES()
    .domain(domain)
    .is_build()
    .app_id(app_id)
    .sort('version', desc=True)
    .is_released()
    .source(['_id', 'version'])
    .filter(
        OR(
            range_filter('last_released', lte=before_date),
            range_filter('built_on', lte=before_date)
        )
    )
    .size(1)
)
```

**Review by mjriley — COMMENTED** [review]
Looks good. Mostly some minor suggestions that we could do to clarify some of the code, but do have some questions over whether some optimizations are necesssary

File: corehq/apps/enterprise/enterprise.py (hunk 10 above)
Comment (gherceg) [reply corehq/apps/enterprise/enterprise.py]: Jing and I offlined about this and I believe cleared this one up. Thanks for doing the exploration on this and sorry that it didn't help anything 😅

File: corehq/apps/enterprise/enterprise.py [hunk 14]
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
...
+        :param app_id: The application id
+        :param time: A datetime object representing the date of the build version to compare against
+        :return: The latest build version available at the given date
+        """
+        builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+        latest_build = self._find_latest_build_version_from_builds(builds, time)
+
+        if latest_build is None:
+            builds = self.get_app_builds(domain, app_id, start=self.INITIAL_QUERY_LIMIT)
+            latest_build = self._find_latest_build_version_from_builds(builds, time)
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: I did this experiment:
image
I plan to change Initial Query Limit to 15. 
Where should I document it?

[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: But since we are not displaying a metric in front end. I guess I can remove that logic too. 08443df3c5703913c93a4a71b0f007bc9f919881

File: corehq/apps/enterprise/enterprise.py [hunk 15]
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
...
+                    latest_version,
+                    build['build_version'],
+                    self.format_date(DateTimeProperty.deserialize(build['build_version_date'])),
+                ])
+
+        return rows
+
+    def total_for_domain(self, domain):
+        app_ids = get_app_ids_in_domain(domain)
+        total_last_builds = 0
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/enterprise/enterprise.py:705]: looks like this was updated in a different location

File: corehq/apps/enterprise/enterprise.py [hunk 16]
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
...
+
+    def get_latest_build_version_at_time(self, domain, app_id, time):
+        """
+        Get the latest build version available at the given time.
+
+        :param domain: The domain of the app
+        :param app_id: The application id
+        :param time: A datetime object representing the date of the build version to compare against
+        :return: The latest build version available at the given date
+        """
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/enterprise/enterprise.py:746]: from review: rename to `get_latest_build_version`, and rename last parameter to `at_datetime` and remove docstring? unless folks have other ideas.

File: corehq/apps/enterprise/enterprise.py [hunk 17]
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
...
+        return self.builds_by_app_id[app_id]
+
+    def _find_latest_build_version_from_builds(self, all_builds, time):
+        """
+        Get the latest build version at the time
+
+        :param all_builds: List of raw build documents sorted by version in descending order
+        :param time: A datetime object representing the date of the build version to compare against
+        :return: The latest build version available at the given date.
+        """
```
[the remark this answers is not in the document]
Comment (biyeun) [reply corehq/apps/enterprise/enterprise.py:773]: from review: rename `time` to `at_datetime`, remove doctstring. add tests

File: corehq/apps/enterprise/enterprise.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: This is removed in 12a4ff624b06a8581092cf70b8c11f085340e6b4 since we're not displaying metric

File: corehq/apps/enterprise/enterprise.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: 5bb7b5a94a024e004b718c4399b9da7892252b93

File: corehq/apps/enterprise/enterprise.py [hunk 18]
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
...
+            else:
+                # last_released is added in 2019, build before 2019 don't have this field
+                # TODO: have a migration to populate last_released from built_on
+                # Then this code can be modified to use last_released only
+                released_date = build_doc['last_released'] or build_doc['built_on']
+                build_info = {
+                    'version': build_doc['version'],
+                    'last_released': DateTimeProperty.deserialize(released_date)
+                }
+                self.build_by_build_id[build_id] = build_info
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: 8a4accd4b46b7b5cd4469ad4ea4eaea26bfa9bd0

File: corehq/apps/enterprise/enterprise.py [hunk 19]
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
...
+                .sort('version', desc=True)
+                .is_released()
+                .source(['_id', 'version', 'last_released', 'built_on'])
+                .start(start)
+            )
+            if limit:
+                app_es = app_es.size(limit)
+            self.builds_by_app_id[app_id] = app_es.run().hits
+
+        return self.builds_by_app_id[app_id]
```
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: c79f198120fb8001c67bc05d5274c5f2a35deace

File: corehq/apps/enterprise/enterprise.py [hunk 20]
```diff
@@ -666,51 +665,31 @@ def rows(self):
...
-
-        for domain in self.account.get_domains():
-            domain_out_of_date_rows, domain_last_builds = self.total_for_domain(domain)
-            total_out_of_date_rows += domain_out_of_date_rows
-            total_last_builds += domain_last_builds
-
-        total_up_to_date = total_last_builds - total_out_of_date_rows
-
-        return _format_percentage_for_enterprise_tile(total_up_to_date, total_last_builds)
+        return '--'
```
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:668]: Can we add a comment explaining why we're skipping this display for now?

File: corehq/apps/enterprise/enterprise.py [hunk 21]
```diff
@@ -727,73 +706,57 @@ def _get_user_builds(self, domain, app_ids):
...
+        if not build_info:
+            # last_released is added in 2019, build before 2019 don't have this field
+            # TODO: have a migration to populate last_released from built_on
+            # Then this code can be modified to use last_released only
+            released_date = build_doc.get('last_released') or build_doc['built_on']
+            build_info = {
+                'version': build_doc['version'],
+                'last_released': DateTimeProperty.deserialize(released_date)
+            }
+            self.build_info_cache[build_id] = build_info
```
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:757]: Can you explain the role of `build_info_cache` here? It looks like we're already caching the raw results from the elasticsearch query. Then, within `_find_latest_build_version_from_builds`, we just want to turn that raw doc into  something we can use to determine whether or not this build's release date was prior to the datetime specified. A conversion function doesn't sound like it would need a cache. Was this something that came up while profiling? If we are caching these results to avoid repeated calls to `DateTimeProperty.deserialize`, perhaps that is something we can do when we first cache the elasticsearch results?

File: corehq/apps/enterprise/enterprise.py [hunk 22]
```diff
@@ -635,3 +634,168 @@ def rows_for_domain(self, domain_obj):
...
+    def rows_for_domain(self, domain):
+        rows = []
+        app_name_by_id = {}
+        app_ids = get_app_ids_in_domain(domain)
+
+        for user, build, latest_version in self._get_user_builds(domain, app_ids):
+            if is_out_of_date(str(build['build_version']), str(latest_version)):
+                app_id = build['app_id']
+                if app_id not in app_name_by_id:
+                    app_name_by_id[app_id] = Application.get_db().get(app_id).get('name')
```
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:681]: This definitely seems like the performance vs readability discussion. Now that it appears we no longer have the strict performance concerns (due to not populating the summary data), do you think we should emphasize the readability/maintainability side rather than the 20% performance gain?

File: corehq/apps/enterprise/enterprise.py [hunk 23]
```diff
@@ -666,51 +665,31 @@ def rows(self):
...
         app_ids = get_app_ids_in_domain(domain)
 
-        for user, build, latest_version in self._get_user_builds(domain, app_ids):
-            if is_out_of_date(str(build['build_version']), str(latest_version)):
-                app_id = build['app_id']
+        for row_data in self._get_user_builds(domain, app_ids):
+            version_in_use = str(row_data['build']['build_version'])
+            latest_version = str(row_data['latest_version'])
+            if is_out_of_date(version_in_use, latest_version):
+                app_id = row_data['build']['app_id']
```
Comment (mjriley) [inline corehq/apps/enterprise/enterprise.py:679]: I don't know if this is something you need to act on, but wanted to bring it up -- I think I struggled to figure out what was going on here, because `_get_user_builds` doesn't sound like it's going to be returning latest version information, and because `row_data` doesn't really suggest what is being dealt with.

In pseudo-code, I'd imagine this is what we're trying to do:
```
retrieve app-last-usage information for all users in a domain
for each app-last-usage:
  retrieve the most current version at the time the app was last used
  if that version is not the same version as is in our information, then
    add a row to our report with the relevant details
```

I'm not sure the above is any better, but if you feel there's some confusion with what is going on with `row_data`, perhaps its worth addressing

**Review by mjriley — COMMENTED** [review]
I think I have some remaining questions on our performance tradeoffs and some of the caching

File: corehq/apps/enterprise/enterprise.py (hunk 20 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:669]: 3a0985a9bbb38580f6bf7dd01267f0902ddd6478

File: corehq/apps/enterprise/enterprise.py (hunk 21 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:758]: `build_info_cache` is to cache all required information of a build, mainly version number and last released date. I did this because I realized deserialize also have an impact on performance, so I want to do less deserialization.
> perhaps that is something we can do when we first cache the elasticsearch results?

This will result in deserialize unnecessary build, for example, if an app already have 1000 builds, then it is very likely the first 500 build's last released date will never be required in the comparison then no need to deserialize.

File: corehq/apps/enterprise/enterprise.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:682]: Can you explain more on why `Application.get_db()` have readability issue?

File: corehq/apps/enterprise/enterprise.py (hunk 23 above)
Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:680]: I think I rename it to `row_data` because `build['build']['build_version]` looks weird. What do you think of change line 675 to `for build_and_latest_version in self.all_last_builds_with_latest_version(domain, app_ids):
`

File: corehq/apps/enterprise/enterprise.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:794]: This is the Law of Demeter, that I linked above. When modules start knowing about more than 1 level of internal structure of another module, things tend to get brittle -- unexpected errors crop up and the code is more difficult to maintain. I'm sure the link I provided above has a more full description. `Application.get_db().get(app_id).get('name')` means that this enterprise module essentially knows about the data access layer. Not just how to query the database (`Application.get_db().get(app_id)`) but also about how the database internally represents an Application. The application layer only generally wants to deal with fully constructed `Application` objects, which gives us flexibility -- `name` could be stored in an arbitrary manner in the database, but anyone interacting with the `Application` object wouldn't need to know those details. With this code, we do.

File: corehq/apps/enterprise/enterprise.py (hunk 23 above)
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:792]: That seems like a good improvement

Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py]: Thanks! df144d43d8404bb056f1f5370cf4176ca312f11a

File: corehq/apps/enterprise/enterprise.py (hunk 21 above)
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:874]: If you haven't seen it yet, there's an infamous statement about ["Premature optimization is the root of all evil"](https://wiki.c2.com/?PrematureOptimization). 

I wrote up a quick performance test to gauge how big of an impact deserialization had:
```
from datetime import datetime
from dimagi.utils.dates import safe_strftime
from dimagi.utils.parsing import ISO_DATETIME_FORMAT
from dimagi.ext.couchdbkit import DateTimeProperty
import random

TIMESTAMP_FLOOR = 1_000_000_000
TIMESTAMP_CEIL = 1_730_000_000

def create_datetime_string():
    return safe_strftime(
        datetime.fromtimestamp(
            random.randrange(TIMESTAMP_FLOOR, TIMESTAMP_CEIL)
        ), ISO_DATETIME_FORMAT)

def create_n_dates(n):
    return [create_datetime_string() for i in range(n)]

def time_deserialize():
    num_elements = 1000
    all_dates = create_n_dates(num_elements)
    deserialized_dates = [None] * num_elements

    start = datetime.now()
    for i in range(num_elements):
        deserialized_dates[i] = DateTimeProperty.deserialize(all_dates[i])
    end = datetime.now()

    print(f'Took: {end - start}')

```

TL;DR -- the time to deserialize 1K datestrings, on my machine, took 10K to 15K microseconds -- i.e. milliseconds.

From the current code, it appears `build_info_cache` only cares about the build number and last released date -- I don't see any other data being used. Given that deserialization has virtually no performance impact, I think it makes the code cleaner for `get_app_builds` to return only the needed information (version, last_released_date) and do the deserialization immediately. This allows us to get rid of `build_info_cache`

**Review by mjriley — APPROVED** [review]
I'm approving to allow this tile to be live on production. I'm convinced this code works, but I hope we can clean up this code in the future to make it more readable and maintainable.

File: corehq/apps/enterprise/enterprise.py (hunk 21 above)
Comment (mjriley) [reply corehq/apps/enterprise/enterprise.py:874]: If you haven't seen it yet, there's an infamous statement about ["Premature optimization is the root of all evil"](https://wiki.c2.com/?PrematureOptimization). 

I wrote up a quick performance test to gauge how big of an impact deserialization had:
```
from datetime import datetime
from dimagi.utils.dates import safe_strftime
from dimagi.utils.parsing import ISO_DATETIME_FORMAT
from dimagi.ext.couchdbkit import DateTimeProperty
import random

TIMESTAMP_FLOOR = 1_000_000_000
TIMESTAMP_CEIL = 1_730_000_000

def create_datetime_string():
    return safe_strftime(
        datetime.fromtimestamp(
            random.randrange(TIMESTAMP_FLOOR, TIMESTAMP_CEIL)
        ), ISO_DATETIME_FORMAT)

def create_n_dates(n):
    return [create_datetime_string() for i in range(n)]

def time_deserialize():
    num_elements = 1000
    all_dates = create_n_dates(num_elements)
    deserialized_dates = [None] * num_elements

    start = datetime.now()
    for i in range(num_elements):
        deserialized_dates[i] = DateTimeProperty.deserialize(all_dates[i])
    end = datetime.now()

    print(f'Took: {end - start}')

```

TL;DR -- the time to deserialize 1K datestrings, on my machine, took 10K to 15K microseconds -- i.e. milliseconds.

From the current code, it appears `build_info_cache` only cares about the build number and last released date -- I don't see any other data being used. Given that deserialization has virtually no performance impact, I think it makes the code cleaner for `get_app_builds` to return

Comment (jingcheng16) [reply corehq/apps/enterprise/enterprise.py:874]: I finally find the commit where I decide to deserialize separately without wrapping the raw document: https://github.com/dimagi/commcare-hq/commit/f94b4b31014a3b7da4c709e1361746aa534fc7e3
I think because I cached the wrapped result initially, so when I change to only deserialize the datetime without wrapping it, I never thought of getting rid of the cache. 
I will put up a PR to remove the cache.
